distributed/scheduler.py

```diff
@@ -8408,3 +8452,50 @@ def transition(
                 self.metadata[key] = ts.metadata
                 self.state[key] = finish
                 self.keys.discard(key)
+
+
+def coassignmnet_groups(
+    tasks: Sequence[TaskState], start: int = 0
+) -> dict[int, set[TaskState]]:
+    groups = {}
+    group = start
+    ix = 0
+    min_prio = None
+    max_prio = None
+
+    group_dependents_seen = set()
+    while ix < len(tasks):
+        current = tasks[ix]
+        if min_prio is None:
+            min_prio = ix
+
+        if current in group_dependents_seen or not current.dependents:
+            min_prio = None
+            max_prio = None
+            ix += 1
+            continue
+        # There is a way to implement this faster by just continuing to iterate
+        # over ix and check if the next is a dependent or not. I chose to go
+        # this route because this is what we wrote down initially
+        next = min(current.dependents, key=lambda ts: ts.priority)
+        next_ix = tasks.index(next)
+
+        # Detect a jump
+        if next_ix != ix + 1:
+            while len(next.dependents) == 1:
+                dep = list(next.dependents)[0]
+                if len(dep.dependencies) != 1:
+                    # This algorithm has the shortcoming that groups may grow too large if the dependent of a group
+                    group_dependents_seen.add(dep)
+                    break
+                next = dep
```

Comment:
Two things where this deviates from the original whiteboard implementation
1. I ended up walking linear chains after all. This may not be necessary after 2.) any more, I haven't checked.
2. I'm breaking early by excluding any dependents of groups. This is a but ugly but pragmatic.